src/Database/LSMTree/Internal/WriteBufferReader.hs

```diff
@@ -0,0 +1,155 @@
+import qualified Database.LSMTree.Internal.WriteBuffer as WB
+import           Database.LSMTree.Internal.WriteBufferBlobs (WriteBufferBlobs)
+import qualified Database.LSMTree.Internal.WriteBufferBlobs as WBB
+import qualified System.FS.API as FS
+import           System.FS.API (HasFS)
+import qualified System.FS.BlockIO.API as FS
+import           System.FS.BlockIO.API (HasBlockIO)
+
+{-# SPECIALISE
+    readWriteBuffer ::
+         TempRegistry IO
+      -> ResolveSerialisedValue
+      -> HasFS IO h
+      -> HasBlockIO IO h
+      -> WriteBufferFsPaths
+      -> IO (WriteBuffer, Ref (WriteBufferBlobs IO h))
+  #-}
+readWriteBuffer ::
+     (MonadMVar m, MonadMask m, MonadSTM m, MonadST m)
+  => TempRegistry m
+  -> ResolveSerialisedValue
+  -> HasFS m h
+  -> HasBlockIO m h
+  -> WriteBufferFsPaths
+  -> m (WriteBuffer, Ref (WriteBufferBlobs m h))
+readWriteBuffer reg resolve hfs hbio wbPaths =
+  bracket (new reg hfs hbio wbPaths) close $ \reader@WriteBufferReader{..} ->
+    (,) <$> readEntries reader <*> pure readerWriteBufferBlobs
+  where
+    readEntries reader = readEntriesAcc WB.empty
+      where
+        readEntriesAcc acc = next reader >>= \case
+          Empty -> pure acc
+          ReadEntry key entry -> readEntriesAcc $ WB.addEntry resolve key (rawBlobRefSpan <$> toFullEntry entry) acc
+
+-- | Allows reading the k\/ops of a run incrementally, using its own read-only
+-- file handle and in-memory cache of the current disk page.
+--
+-- New pages are loaded when trying to read their first entry.
+data WriteBufferReader m h = WriteBufferReader {
+      -- | The disk page currently being read. If it is 'Nothing', the reader
+      -- is considered closed.
+      readerCurrentPage      :: !(MutVar (PrimState m) (Maybe RawPage))
+      -- | The index of the entry to be returned by the next call to 'next'.
+    , readerCurrentEntryNo   :: !(PrimVar (PrimState m) Word16)
+    , readerKOpsHandle       :: !(FS.Handle h)
+    , readerWriteBufferBlobs :: !(Ref (WriteBufferBlobs m h))
+    , readerHasFS            :: !(HasFS m h)
+    , readerHasBlockIO       :: !(HasBlockIO m h)
+    }
+
+{-# SPECIALISE
+    new ::
+         TempRegistry IO
+      -> HasFS IO h
+      -> HasBlockIO IO h
+      -> WriteBufferFsPaths
+      -> IO (WriteBufferReader IO h)
+  #-}
+-- | See 'Database.LSMTree.Internal.RunReader.new'.
+new :: forall m h.
+     (MonadMVar m, MonadST m, MonadMask m)
+  => TempRegistry m
+  -> HasFS m h
+  -> HasBlockIO m h
+  -> WriteBufferFsPaths
+  -> m (WriteBufferReader m h)
+new reg readerHasFS readerHasBlockIO fsPaths = do
+  (readerKOpsHandle :: FS.Handle h) <- FS.hOpen readerHasFS (Paths.writeBufferKOpsPath fsPaths) FS.ReadMode
+  -- Double the file readahead window (only applies to this file descriptor)
+  FS.hAdviseAll readerHasBlockIO readerKOpsHandle FS.AdviceSequential
+  readerWriteBufferBlobs <- allocateTemp reg (WBB.open readerHasFS (Paths.writeBufferBlobPath fsPaths) FS.AllowExisting) releaseRef
```

Comment:
It would be closer to how the `RunReader` works if we are required to pass in a `BlobFile`/`WriteBufferBlobs` instead of creating it here. It would mean that a `WriteBufferReader` behaves the same as a `RunReader` with regards to references: it duplicates input references, but it does not create them